File: src/handler.ts
import { ValidationError } from './error';
import { Request, Response } from 'express';
import { GenerateCommitMessageRequest, CacheType, AIService, Provider } from './types';

export function getHealthzHandler(html: (content: string) => string) {
	const message = 'ok';

	return (req: Request, res: Response) => {
		if (req.get('Content-Type') === 'application/json') {
			return res.status(200).json({ message });
		}

		return res.setHeader('Content-Type', 'text/html').status(200).send(html('ok'));
	};
}

export function getInstallDotShHandler(
	fs: typeof import('node:fs/promises'),
	cache: CacheType,
	installDotSh: string,
	installDotShPath: string,
	html: (content: string) => string,
	extractDomain: (req: Request) => string,
) {
	let file = cache.get(installDotShPath);

	let domain = cache.get('domain');

	const message = `Run this command from your terminal:`;

	const template = (command: string) => html(`${message} <span class="command">${command}</span>`);

	return async (req: Request, res: Response) => {
		if (!domain) {
			domain = extractDomain(req);
			cache.set('domain', domain);
		}

		if (!req.headers['user-agent']?.includes('curl')) {
			const command = `curl -s ${domain}/install.sh | sh`;

			if (req.get('Content-Type') === 'application/json') {
				return res.status(200).json({ message: `${message} ${command}` });
			}

			return res.setHeader('Content-Type', 'text/html').status(200).send(template(command));
		}

		if (!file) {
			file = await fs.readFile(installDotShPath, 'utf-8');
			cache.set(installDotShPath, file);
		}

		return res
			.setHeader('Content-Disposition', `attachment; filename=${installDotSh}`)
			.setHeader('Cache-Control', 'public, max-age=2592000') // Cache for 30 days
			.status(200)
			.send(file);
	};
}

export function getIndexHandler(
	fs: typeof import('node:fs/promises'),
	cache: CacheType,
	commitDotSh: string,
	commitDotShPath: string,
	html: (content: string) => string,
	extractDomain: (req: Request) => string,
) {
	let file = cache.get(commitDotShPath);

	let domain = cache.get('domain');

	const message = `Run this command from your terminal:`;

	const template = (command: string) => html(`${message} <span class="command">${command}</span>`);

	return async (req: Request, res: Response) => {
		if (!domain) {
			domain = extractDomain(req);
			cache.set('domain', domain);
		}

		if (!req.headers['user-agent']?.includes('curl')) {
			const command = `curl -s ${domain}/ | sh -- -k 'YOUR_OPEN_AI_API_KEY'`;

			if (req.get('Content-Type') === 'application/json') {
				return res.status(200).json({ message: `${message} ${command}` });
			}

			return res.setHeader('Content-Type', 'text/html').status(200).send(template(command));
		}

		if (!file) {
			file = await fs.readFile(commitDotShPath, 'utf-8');
			file = file.replace(/http:\/\/localhost/g, domain + '/');
			cache.set(commitDotShPath, file);
		}

		return res
			.setHeader('Content-Disposition', `attachment; filename=${commitDotSh}`)
			.setHeader('Cache-Control', 'public, max-age=2592000') // Cache for 30 days
			.status(200)
			.send(file);
	};
}

export function postGenerateCommitMessageHandler(ai: (type?: Provider) => AIService) {
	return async (req: GenerateCommitMessageRequest, res: Response) => {
		const { diff, provider, apiKey } = req.body;

		if (!diff || !diff.trim().length) {
			throw new ValidationError('Diff must not be empty!');
		}

		if (provider && provider !== 'openai' && provider !== 'claudeai') {
			throw new ValidationError('Invalid provider specified!');
		}

		const message = await ai(provider).generate(diff, apiKey);

		return res.status(200).json({ message });
	};
}

File: cmd/handler.go
package main

import (
	"bytes"
	"fmt"
	"html/template"
	"io"
	"net/http"
	"strings"

	"github.com/wajeht/commit/assets"
)

var (
	homeTemplate    = pageTemplate("templates/index.html")
	installTemplate = pageTemplate("templates/install.html")
)

type pageData struct {
	Title   string
	Domain  string
	Command string
}

func pageTemplate(page string) *template.Template {
	return template.Must(template.ParseFS(assets.Embeddedfiles, "templates/base.html", page))
}

func (app *application) handleHealthz(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/plain")
	w.WriteHeader(http.StatusOK)
	w.Write([]byte("ok"))
}

func (app *application) handleFavicon(w http.ResponseWriter, r *http.Request) {
	file, err := assets.Embeddedfiles.Open("static/favicon.ico")
	if err != nil {
		app.serverError(w, r, err)
		return
	}
	defer file.Close()

	w.Header().Set("Content-Type", "image/x-icon")
	if _, err := io.Copy(w, file); err != nil {
		app.reportServerError(r, err)
	}
}

func (app *application) handleRobotsTxt(w http.ResponseWriter, r *http.Request) {
	file, err := assets.Embeddedfiles.Open("static/robots.txt")
	if err != nil {
		app.serverError(w, r, err)
		return
	}
	defer file.Close()

	w.Header().Set("Content-Type", "text/plain")
	if _, err := io.Copy(w, file); err != nil {
		app.reportServerError(r, err)
	}
}

func (app *application) handleInstallSh(w http.ResponseWriter, r *http.Request) {
	domain := app.domain(r)

	userAgent := r.Header.Get("User-Agent")
	isCurl := strings.Contains(userAgent, "curl")

	if !isCurl {
		command := fmt.Sprintf("curl -s %s/install.sh | bash", domain)
		message := "Run this command from your terminal:"
		accept := r.Header.Get("Accept")

		if strings.Contains(accept, "application/json") {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusOK)
			fmt.Fprintf(w, `{"message":"%s %s"}`, message, command)
			return
		}

		var page bytes.Buffer
		if err := installTemplate.ExecuteTemplate(&page, "base.html", pageData{
			Title:   "Install Commit",
			Domain:  domain,
			Command: command,
		}); err != nil {
			app.serverError(w, r, err)
			return
		}

		w.Header().Set("Content-Type", "text/html; charset=utf-8")
		w.WriteHeader(http.StatusOK)
		if _, err := page.WriteTo(w); err != nil {
			app.reportServerError(r, err)
		}
		return
	}

	file, err := assets.Embeddedfiles.Open("sh/install.sh")
	if err != nil {
		app.serverError(w, r, err)
		return
	}
	defer file.Close()

	w.Header().Set("Content-Type", "text/plain")
	w.Header().Set("Content-Disposition", "attachment; filename=install.sh")
	w.Header().Set("Cache-Control", "public, max-age=2592000")
	w.WriteHeader(http.StatusOK)
	if _, err := io.Copy(w, file); err != nil {
		app.reportServerError(r, err)
	}
}

func (app *application) handleHome(w http.ResponseWriter, r *http.Request) {
	if r.URL.Path != "/" {
		app.notFound(w, r)
		return
	}

	domain := app.domain(r)

	userAgent := r.Header.Get("User-Agent")
	isCurl := strings.Contains(userAgent, "curl")

	if !isCurl {
		command := fmt.Sprintf("curl -s %s | bash", domain)
		message := "Run this command from your terminal:"
		accept := r.Header.Get("Accept")

		if strings.Contains(accept, "application/json") {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusOK)
			fmt.Fprintf(w, `{"message":"%s %s"}`, message, command)
			return
		}

		var page bytes.Buffer
		if err := homeTemplate.ExecuteTemplate(&page, "base.html", pageData{
			Title:  "Commit",
			Domain: domain,
		}); err != nil {
			app.serverError(w, r, err)
			return
		}

		w.Header().Set("Content-Type", "text/html; charset=utf-8")
		w.WriteHeader(http.StatusOK)
		if _, err := page.WriteTo(w); err != nil {
			app.reportServerError(r, err)
		}
		return
	}

	file, err := assets.Embeddedfiles.Open("sh/commit.sh")
	if err != nil {
		app.serverError(w, r, err)
		return
	}
	defer file.Close()

	content, err := io.ReadAll(file)
	if err != nil {
		app.serverError(w, r, err)
		return
	}

	modifiedContent := strings.ReplaceAll(string(content), "http://localhost", domain)

	w.Header().Set("Content-Type", "text/plain")
	w.Header().Set("Content-Disposition", "attachment; filename=commit.sh")
	w.Header().Set("Cache-Control", "public, max-age=2592000")
	w.WriteHeader(http.StatusOK)
	if _, err := w.Write([]byte(modifiedContent)); err != nil {
		app.reportServerError(r, err)
	}
}
